llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -12987,6 +12989,225 @@ struct AAAllocationInfoCallSiteArgument : AAAllocationInfoImpl {
 };
 } // namespace
 
+/// ----------- AAConvertOutArgument ----------
+namespace {
+static bool isEligibleArgument(const Argument &Arg, Attributor &A,
+                               const AbstractAttribute &AA) {
+  if (!Arg.getType()->isPointerTy())
+    return false;
+
+  const IRPosition &ArgPos = IRPosition::argument(Arg);
+  auto *AAMem = A.getAAFor<AAMemoryBehavior>(AA, ArgPos, DepClassTy::OPTIONAL);
+  auto *NoAlias = A.getAAFor<AANoAlias>(AA, ArgPos, DepClassTy::OPTIONAL);
+
+  return AAMem && NoAlias && AAMem->isAssumedWriteOnly() &&
+         NoAlias->isAssumedNoAlias() && !Arg.hasPointeeInMemoryValueAttr();
+}
+
+struct AAConvertOutArgumentFunction final : AAConvertOutArgument {
+  AAConvertOutArgumentFunction(const IRPosition &IRP, Attributor &A)
+      : AAConvertOutArgument(IRP, A) {}
+
+  SmallVector<bool> ArgumentsStates;
+
+  /// See AbstractAttribute::updateImpl(...).
+  void initialize(Attributor &A) override {
+    const Function *F = getAssociatedFunction();
+    if (!F || F->isDeclaration())
+      return;
+
+    // Assume that all  args are convertable at the begining.
+    ArgumentsStates.resize(F->arg_size(), true);
+  }
+
+  /// See AbstractAttribute::updateImpl(...).
+  ChangeStatus updateImpl(Attributor &A) override {
+    const Function *F = getAssociatedFunction();
+    if (!F || F->isDeclaration())
+      return indicatePessimisticFixpoint();
+
+    auto NewStates = ArgumentsStates;
+    for (unsigned ArgIdx = 0; ArgIdx < F->arg_size(); ++ArgIdx)
+      if (!isEligibleArgument(*F->getArg(ArgIdx), A, *this))
+        NewStates[ArgIdx] = false;
+
+    bool Changed = NewStates == ArgumentsStates;
+    ArgumentsStates = NewStates;
+    return Changed ? ChangeStatus::CHANGED : ChangeStatus::UNCHANGED;
+  }
+
+  /// See AbstractAttribute::manifest(...).
+  ChangeStatus manifest(Attributor &A) override {
+    Function &F = *getAssociatedFunction();
+    DenseMap<Argument *, Type *> PtrToType;
+    SmallVector<Argument *, 4> CandidateArgs;
+
+    for (unsigned ArgIdx = 0; ArgIdx < F.arg_size(); ++ArgIdx) {
+      Argument *Arg = F.getArg(ArgIdx);
+      if (!isEligibleArgument(*Arg, A, *this))
+        continue;
+
+      CandidateArgs.push_back(Arg);
+      // AAPointerInfo on args
+      for (auto &Use : Arg->uses())
+        if (auto *Store = dyn_cast<StoreInst>(Use.getUser()))
+          PtrToType[Arg] = Store->getValueOperand()->getType();
```

Comment:
You need to handle more than just stores users here. What if the user is a gep, or some other instruction. You can use AAPointerInfo instead of AAMemoryBehavior and then iterate over all the instructions that actually modify the pointer. It will also tell you the offsets, which you'll likely need.